src/api/procedures/addInstruction.ts

```diff
@@ -98,128 +118,198 @@ export async function prepareAddInstruction(
-    legs.map(async ({ from, to, amount, token }) => {
-      const fromId = portfolioLikeToPortfolioId(from);
-      const toId = portfolioLikeToPortfolioId(to);
-
-      await Promise.all([
-        assertPortfolioExists(fromId, context),
-        assertPortfolioExists(toId, context),
-      ]);
-
-      const rawFromPortfolio = portfolioIdToMeshPortfolioId(fromId, context);
-      const rawToPortfolio = portfolioIdToMeshPortfolioId(toId, context);
-
-      rawLegs.push({
-        from: rawFromPortfolio,
-        to: rawToPortfolio,
-        asset: stringToTicker(typeof token === 'string' ? token : token.ticker, context),
-        amount: numberToBalance(amount, context),
-      });
-    })
-  );
-
-  let newInstruction;
+  let affirmedInstructions: PostTransactionValue<Instruction[]> | undefined;
+  let addedInstructions: PostTransactionValue<Instruction[]>;
 
-  if (portfoliosToAffirm.length) {
-    [newInstruction] = this.addTransaction(
+  if (addAndAffirmInstructions.length) {
+    [affirmedInstructions] = this.addBatchTransaction(
       settlement.addAndAffirmInstruction,
       {
         resolvers: [createAddInstructionResolver(context)],
       },
-      rawVenueId,
-      rawSettlementType,
-      rawTradeDate,
-      rawValueDate,
-      rawLegs,
-      portfoliosToAffirm.map(portfolio =>
-        portfolioIdToMeshPortfolioId(portfolioLikeToPortfolioId(portfolio), context)
-      )
+      addAndAffirmInstructions
     );
-  } else {
-    [newInstruction] = this.addTransaction(
+  }
+
+  if (addInstructions.length) {
+    [addedInstructions] = this.addBatchTransaction(
       settlement.addInstruction,
       {
-        resolvers: [createAddInstructionResolver(context)],
+        resolvers: [createAddInstructionResolver(context, affirmedInstructions)],
       },
-      rawVenueId,
-      rawSettlementType,
-      rawTradeDate,
-      rawValueDate,
-      rawLegs
+      addInstructions
     );
+
+    affirmedInstructions = addedInstructions;
   }
 
-  return newInstruction;
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  return affirmedInstructions!;
 }
 
 /**
  * @hidden
  */
 export async function getAuthorization(
-  this: Procedure<Params, Instruction, Storage>,
+  this: Procedure<Params, Instruction[], Storage>,
   { venueId }: Params
 ): Promise<ProcedureAuthorization> {
   const {
     storage: { portfoliosToAffirm },
   } = this;
-  let transactions: TxTag[];
 
-  if (portfoliosToAffirm.length) {
-    transactions = [TxTags.settlement.AddAndAffirmInstruction];
-  } else {
-    transactions = [TxTags.settlement.AddInstruction];
-  }
+  const portfolios: (DefaultPortfolio | NumberedPortfolio)[] = [];
+  const transactions: SettlementTx[] = [];
+
+  portfoliosToAffirm.forEach(portfoliosList => {
+    transactions.push(
+      portfoliosList.length
+        ? TxTags.settlement.AddAndAffirmInstruction
+        : TxTags.settlement.AddInstruction
+    );
+    portfolios.push(...portfoliosList);
```

Comment:
You can use lodash `union` for transactions and `unionWith` together with `isEqual` for portfolios so that you don't have to use `uniq` later on a big array